rundeckapp/grails-app/services/rundeck/services/asyncimport/AsyncImportService.groovy

```diff
@@ -0,0 +1,1073 @@
+import com.dtolabs.rundeck.core.common.IRundeckProject
+import grails.converters.JSON
+import grails.events.EventPublisher
+import grails.events.annotation.Subscriber
+import groovy.json.JsonSlurper
+import liquibase.logging.LogFactory
+import org.apache.commons.io.FileUtils
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import rundeck.services.ConfigurationService
+import rundeck.services.FrameworkService
+import rundeck.services.ProjectService
+
+import java.nio.charset.StandardCharsets
+import java.nio.file.FileVisitOption
+import java.nio.file.Files
+import java.nio.file.NoSuchFileException
+import java.nio.file.Path
+import java.nio.file.Paths
+import java.nio.file.StandardCopyOption
+import java.util.function.Predicate
+import java.util.stream.Collectors
+import java.util.zip.ZipEntry
+import java.util.zip.ZipInputStream
+import java.util.zip.ZipOutputStream
+
+/**
+ * Service that handles all the Asynchronous import functionality.
+ *
+ */
+class AsyncImportService implements AsyncImportStatusFileOperations, EventPublisher {
+
+    FrameworkService frameworkService
+    ProjectService projectService
+    ConfigurationService configurationService
+    Logger logger = LoggerFactory.getLogger(this.class)
+
+    // Constants
+    static final String TEMP_DIR = stripSlashFromString(System.getProperty("java.io.tmpdir"))
+    static final Path BASE_WORKING_DIR = Paths.get(TEMP_DIR + File.separator + "AImport-WD-")
+    static final String DISTRIBUTED_EXECUTIONS_FILENAME = "distributed_executions"
+    static final String TEMP_PROJECT_SUFFIX = 'AImportTMP-'
+    static final String JSON_FILE_PREFIX = 'AImport-status-'
+    static final String JSON_FILE_EXT = '.json'
+    static final String EXECUTION_DIR_NAME = 'executions'
+    static final String MODEL_PROJECT_NAME_SUFFIX = 'rundeck-model-project'
+    static final String MODEL_PROJECT_NAME_EXT = '.jar'
+    static final String MODEL_PROJECT_INTERNAL_PREFIX = 'rundeck-'
+    static final String MAX_EXECS_PER_DIR_PROP_NAME = "asyncImportConfig.maxDistributedExecutions"
+    static final int KILOBYTE=1024
+
+    static final String EXECUTION_FILE_PREFIX = 'execution-'
+    static final String EXECUTION_FILE_EXT = '.xml'
+    static final String OUTPUT_FILE_PREFIX = 'output-'
+    static final String OUTPUT_FILE_EXT = '.rdlog'
+    static final String STATE_FILE_PREFIX = 'state-'
+    static final String STATE_FILE_EXT = '.state.json'
+
+    /**
+     *
+     * Creates the status file that will be the main report to inform about the whole process,
+     * a project cannot have more than a single status file.
+     *
+     * The file will be stored in project's file storage (Storage table in db)
+     *
+     * @param projectName - Required param to project.
+     *
+     * @return Boolean - "true" if the status file is created.
+     */
+    @Override
+    Boolean createStatusFile(String projectName) {
+        try {
+            if( !statusFileExists(projectName) ){
```

Comment:
why would the status file exist? and what would happen if that is so?

Reply:
Its a layer of validation just to delegate that in other method and prevent a thick save method